etl/src/v2/replication/table_sync.rs

```diff
@@ -1,28 +1,168 @@
+    state_store: S,
+    destination: D,
+    shutdown_rx: watch::Receiver<()>,
+) -> Result<TableSyncResult, TableSyncError>
+where
+    S: StateStore + Clone + Send + 'static,
     D: Destination + Clone + Send + 'static,
 {
-    // 1. Load the state from the TableSyncWorkerState (must be done there since we know that the
-    //    in-memory state is consistent from this point, if we re-read from the store, we already
-    //    gave out the state which could have been changed already).
-    // 2. Check the state and exit if SyncDone, Ready, Unknown
-    // 3. Compute the slot name
-    // 4. Make sure the state is either Init, DataSync, FinishedCopy
-    // 5. Handle slot deletion in case it's in DataSync
-    // 6. Start copy table with transaction
-    // 7. Mark the table as FinishedCopy
+    let inner = table_sync_worker_state.get_inner().read().await;
+
+    let table_id = inner.table_id();
+    let phase_type = inner.replication_phase().as_type();
+
+    // In case the work for this table has been already done, we don't want to continue and we
+    // successfully return.
+    if phase_type == TableReplicationPhaseType::SyncDone
+        || phase_type == TableReplicationPhaseType::Ready
+        || phase_type == TableReplicationPhaseType::Unknown
+    {
+        return Ok(TableSyncResult::SyncNotRequired);
+    }
+
+    // In case the phase is different from the standard phases in which a table sync worker can perform
+    // table syncing, we want to return an error.
+    if !(phase_type == TableReplicationPhaseType::Init
+        || phase_type == TableReplicationPhaseType::DataSync
+        || phase_type == TableReplicationPhaseType::FinishedCopy)
+    {
+        return Err(TableSyncError::InvalidPhase(phase_type));
+    }
+
+    // We are safe to unlock the state here, since we know that the state will be changed by the
+    // apply worker only if `SyncWait` is set, which is not the case if we arrive here, so we are
+    // good to reduce the length of the critical section.
+    drop(inner);
+
+    let slot_name = get_slot_name(&identity, SlotUsage::TableSyncWorker { table_id })?;
+
+    // If we hit this condition, it means that we crashed either before or after finishing the table
+    // copying or even during catchup. Since we don't make any time assumptions about when this worker
+    // is restarted with this state, we need to delete the existing slot (if any) since we might have
+    // lost WAL entries so it's not safe to try and catchup again.
+    if phase_type == TableReplicationPhaseType::DataSync
+        || phase_type == TableReplicationPhaseType::FinishedCopy
+    {
+        if let Err(err) = replication_client.delete_slot(&slot_name).await {
+            // If the slot is not found, we are safe to continue, for any other error, we bail.
+            let PgReplicationError::SlotNotFound(_) = err else {
+                return Err(err.into());
+            };
+        }
+    }
```

Comment:
Slot should only be deleted when the table sync worker is in the `DataSync` state. If it was in `FinishedCopy` it means it had already completed the initial table copy and can use the slot to catch up. We don't want to do the costly initial table copy again if the worker failed during `DataSync`.

Reply:
Yes totally, this was my mistake, I checked the Postgres code now and missed a `goto` they had. I also thought initially that with finished copy we would be good to just move onto `SyncWait`, especially since the replication slot works fine even when the connection is dropped (assuming the table is copied). Will fix this!